drivers/serial/uart_nrfx_uarte.c

```diff
@@ -320,29 +339,32 @@ static int uarte_nrfx_fifo_read(struct device *dev,
 static void uarte_nrfx_irq_tx_enable(struct device *dev)
 {
 	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
 
 	nrf_uarte_int_enable(uarte, NRF_UARTE_INT_ENDTX_MASK);
+	data->disable_tx_irq = false;
 }
 
 /** Interrupt driven transfer disabling function */
 static void uarte_nrfx_irq_tx_disable(struct device *dev)
 {
-	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
-
-	nrf_uarte_int_disable(uarte, NRF_UARTE_INT_ENDTX_MASK);
-
-	/* If there is nothing to send, driver will save an energy
-	 * when TX is stopped.
-	 */
-	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	/* TX IRQ will be disabled after current transmission is finished */
+	data->disable_tx_irq = true;
 }
 
 /** Interrupt driven transfer ready function */
 static int uarte_nrfx_irq_tx_ready_complete(struct device *dev)
 {
 	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 
-	return nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ENDTX);
+	/* ENDTX flag is always on so that ISR is called when we enable TX IRQ.
+	 * Because of that we have to explicitly check if ENDTX interrupt is
+	 * enabled, otherwise this function would always return true no matter
+	 * what would be the source of interrupt.
```

Comment:
Sorry, I don't get it. This function is supposed to return true exactly when the `NRF_UARTE_EVENT_ENDTX` event is set. And it is not true that this event is always set. It is cleared before a transmission is started, see `uarte_nrfx_fifo_fill`.

Reply:
For the application it is a problem if after disabling interrupts `uart_irq_tx_ready()` returns true.

Reply:
> For the application it is a problem if
> after disabling interrupts `uart_irq_tx_ready()` returns true.

Then perhaps the application should be corrected.
The UART API does not describe such behavior as required or expected, does it?

Reply:
Application assumes correctly that it is allowed to send data with `uarte_nrfx_fifo_fill` and ignore tx interrupt by disabling it. I pretty sure this is not problem of an application.

Reply:
> The UART API does not describe such behavior as required or expected, does it?

It does. If it doesn't that's bug. Indeed, uart_irq_tx_ready() should be gated on on interrupts being enabled. Well, I don't know if there're some apps which can be fooled otherwise, but we can't dismiss that. The issue, there's single ISR for both UART RX and TX (and for errors, about which a few people remember). So, ISR can be called by RX interrupt, and should not be wrongly doing TX actions if TX is effectively disabled (and if TX interrupt is disabled, then effectively TX is disabled).

Reply:
I created an issue (#11535) for this to be clarified.

Reply:
This PR fixes current issues with nrf uarte driver acting differently than nrf uart driver. I agree that API has to be clarified and it should be discussed, but changes in this PR don't affect it and are needed now. Also there is ongoing work on new API #10820 which will eventually make current interrupt driven API obsolete and we should focus on that.

Reply:
I'm +1 on this logic, let me approve.

Reply:
Agreed. Let's go on with this fix for now.